File: tests/sdks/python/fixtures/lowkey_vault_fixture.py
import ssl

import pytest
import requests
from azure.identity import DefaultAzureCredential
from azure.keyvault.secrets import SecretClient
from envilder.infrastructure.azure.azure_key_vault_secret_provider import (
    AzureKeyVaultSecretProvider,
)
from testcontainers.core.container import DockerContainer
from testcontainers.core.waiting_utils import wait_for_logs


class _TrustAllSession(requests.Session):
    def __init__(self):
        super().__init__()
        self.verify = False


@pytest.fixture(scope="session")
def lowkey_vault_container():
    container = (
        DockerContainer("nagyesta/lowkey-vault:7.1.32")
        .with_exposed_ports(8443, 8080)
        .with_env(
            "LOWKEY_ARGS",
            "--server.port=8443 --LOWKEY_VAULT_RELAXED_PORTS=true",
        )
    )
    with container:
        wait_for_logs(container, "Started LowkeyVaultApp")
        yield container


@pytest.fixture(scope="session")
def lowkey_vault_url(lowkey_vault_container):
    host = lowkey_vault_container.get_container_host_ip()
    port = lowkey_vault_container.get_exposed_port(8443)
    return f"https://{host}:{port}"


@pytest.fixture(scope="session")
def lowkey_vault_token_url(lowkey_vault_container):
    host = lowkey_vault_container.get_container_host_ip()
    port = lowkey_vault_container.get_exposed_port(8080)
    return f"http://{host}:{port}" "/metadata/identity/oauth2/token"


@pytest.fixture(scope="session")
def azure_secret_client(
    lowkey_vault_url, lowkey_vault_token_url, monkeypatch_session
):
    import os

    os.environ["IDENTITY_ENDPOINT"] = lowkey_vault_token_url
    os.environ["IDENTITY_HEADER"] = "dummy"

    ssl_context = ssl.create_default_context()
    ssl_context.check_hostname = False
    ssl_context.verify_mode = ssl.CERT_NONE

    client = SecretClient(
        vault_url=lowkey_vault_url,
        credential=DefaultAzureCredential(),
        connection_verify=False,
        verify_challenge_resource=False,
    )

    yield client

    os.environ.pop("IDENTITY_ENDPOINT", None)
    os.environ.pop("IDENTITY_HEADER", None)


@pytest.fixture(scope="session")
def azure_provider(azure_secret_client):
    return AzureKeyVaultSecretProvider(azure_secret_client)


@pytest.fixture(scope="session")
def monkeypatch_session():
    from _pytest.monkeypatch import MonkeyPatch

    m = MonkeyPatch()
    yield m
    m.undo()

File: tests/sdks/dotnet/Fixtures/LowkeyVaultFixture.cs
namespace Envilder.Tests.Fixtures;

using Azure.Core.Pipeline;
using Azure.Identity;
using Azure.Security.KeyVault.Secrets;
using DotNet.Testcontainers.Builders;
using DotNet.Testcontainers.Containers;
using System.Net.Http;

public sealed class LowkeyVaultFixture : IAsyncLifetime
{
    private const int VaultPort = 8443;
    private const int TokenPort = 8080;

    private IContainer _container = null!;

    public string VaultUrl { get; private set; } = null!;

    public SecretClient SecretClient { get; private set; } = null!;

    public async Task InitializeAsync()
    {
        _container = new ContainerBuilder("nagyesta/lowkey-vault:7.1.61")
            .WithName($"lowkey-vault-{Guid.CreateVersion7()}")
            .WithPortBinding(VaultPort, true)
            .WithPortBinding(TokenPort, true)
            .WithEnvironment(
                "LOWKEY_ARGS",
                "--server.port=8443 --LOWKEY_VAULT_RELAXED_PORTS=true")
            .WithWaitStrategy(Wait.ForUnixContainer()
                .UntilExternalTcpPortIsAvailable(VaultPort))
            .Build();

        await _container.StartAsync();

        var host = _container.Hostname;
        var vaultPort = _container.GetMappedPublicPort(VaultPort);
        var tokenPort = _container.GetMappedPublicPort(TokenPort);

        VaultUrl = $"https://{host}:{vaultPort}";

        Environment.SetEnvironmentVariable(
            "IDENTITY_ENDPOINT",
            $"http://{host}:{tokenPort}/metadata/identity/oauth2/token");
        Environment.SetEnvironmentVariable(
            "IDENTITY_HEADER",
            "dummy");

        var httpHandler = new HttpClientHandler
        {
            ServerCertificateCustomValidationCallback =
                HttpClientHandler
                    .DangerousAcceptAnyServerCertificateValidator,
        };

        var clientOptions = new SecretClientOptions(
            SecretClientOptions.ServiceVersion.V7_2)
        {
            Transport =
                new HttpClientTransport(new HttpClient(httpHandler)),
            DisableChallengeResourceVerification = true,
        };

        SecretClient = new SecretClient(
            new Uri(VaultUrl),
            new DefaultAzureCredential(),
            clientOptions);
    }

    public async Task DisposeAsync()
    {
        Environment.SetEnvironmentVariable("IDENTITY_ENDPOINT", null);
        Environment.SetEnvironmentVariable("IDENTITY_HEADER", null);

        await _container.DisposeAsync();
    }
}
